UTF detection when missing Byte Order Mark

jpz opened the pull request:
This change adds heuristic detection of UTF-16 and UTF-32 files when they are missing their byte order marks.  

At present we have no strategy for detecting the format of these files.  Feel free to give feedback on the PR by the way, happy to have other eyes on it.

Note I report these files as UTF-16LE, UTF-16BE,  UTF-32LE and UTF-32BE rather than UTF-16 and UTF-32.  This is justified for the following reasons: it's quite material whether it is little endian or big endian - Python will not decode these files correctly when passed decode("utf-16") - Python assumes endian-ness.  This is less important for the files with the BOMs, as UTF-aware readers will generally inspect this and auto-decode the file, but for things missing the BOM, it's really important to be told the endian-ness in hopes of reading the file.  

We could change the reporting of of files with BOM to inform the endian-ness also, but I've avoided having an opinion on that for the moment.

jpz commented:
Rebased the PR to origin/master.

dan-blanchard commented:
Thanks so much for this contribution! I haven't gotten a chance to run through all of it yet, but from what I've seen so far, it looks good.

> Note I report these files as UTF-16LE, UTF-16BE, UTF-32LE and UTF-32BE rather than UTF-16 and UTF-32.

As discussed in [this comment from an old PR](https://github.com/chardet/chardet/pull/73#issuecomment-145955258) this is entirely the correct *and required* behavior. UTF-16LE and UTF-16BE are what one is required to report when there is no BOM, but they are not allowed if there is a BOM.

dan-blanchard commented on chardet/utf1632prober.py:
```diff
@@ -0,0 +1,137 @@
+# appear to be UTF32LE.  Files looking like ( \0 [notzero] )+ appear
+# may be guessed to be UTF16LE.
+
+class UTF1632Prober(CharSetProber):
```
The way we would usually handle this sort of multi-byte encoding is with a [`CodingStateMachine`](https://github.com/chardet/chardet/blob/master/chardet/codingstatemachine.py#L33) and [`CharDistributionAnalysis`](https://github.com/chardet/chardet/blob/master/chardet/chardistribution.py#L40)-based implementation. You can see the prober we have for UTF-8 (which is obviously single-byte) [here](https://github.com/chardet/chardet/blob/master/chardet/utf8prober.py), which only uses the `CodingStateMachien`, but the approach is close to what we would need to do for UTF-16 and UTF-32. In fact, through some crazy bit of coincidence, we already have [unused state machines for UTF-16LE and UTF-16BE in our code](https://github.com/chardet/chardet/blob/master/chardet/mbcssm.py#L390-L498) that I've never noticed until now. It appears to have been there since version 1.0, but never used.

Anyway, I know it would be a bit of work to switch this over to using `CodingStateMachine` like the UTF-8 prober, so I'm not going to require that to merge this, but if you don't do it, I'll probably end up switching this over to that eventually.

jpz replied:
Ah, this would make sense as it would be cleaner - you could verify the existence of the correct escape sequences - my implementation is more of a quick and dirty heuristic.  

I feel grokking this state machine would take me days, so if you don't mind I may defer this task if that's ok.

dan-blanchard commented on chardet/utf1632prober.py:
```diff
@@ -0,0 +1,137 @@
+    def get_confidence(self):
+        confidence = 0.99
+
+        if self.is_likely_utf16le() or self.is_likely_utf16be() or self.is_likely_utf32le() or self.is_likely_utf32be():
```
I'd really like to see a more mathematical basis for the confidence here, rather than the essentially binary 0.99 or 0.01 we have now. Maybe something based on the difference between `EXPECTED_RATIO` and the values you have in `nonzeros_at_mod` and `zeros_at_mod`? 0.99 is a really high confidence, and I like to reserve it for definite yeses. In fact, calculating this confidence is a lot of what the `CharDistributionAnalysis` class does, which is part of why I think that's a more natural fit for this codec.

jpz replied:
I'd say if we look at even 100 bytes (25 code points) and they are all "0 0 0 x 0 0 0 y 0 0 0 z" etc - our confidence is likely to be sky high this is UTF32.  Likewise for UTF16.  

That was my thinking - I realise it's not particularly rigorous to state this without statistical analysis of many texts however, and maybe there are counter examples I'm not aware of.

One way to improve rigour, would be to verify exceptions to zeros are actually extended characters (surrogate pairs) of UTF16 also.  For instance in UTF16 you'd be looking for quads of (D8-DB) xx (DC-DF) xx to be recognised, e.g. - https://en.m.wikipedia.org/wiki/UTF-16#Examples

There are some other unused ranges in UTF16 and UTF32 also, so the detector could be made to flunk to zero if any of these are detected. 

You're making me think again about correctness of my implementation :)

jpz replied:
I reduced the estimate to 0.85 arbitrarily, as I don't have rationale to set it at 0.99.  Happy to discuss this point further.  

I also did some additional checking for valid UTF16/UTF32 with the escapes which improves robustness.

dan-blanchard commented:
Also, thanks so much for the PR!

jpz commented on test.py:
```diff
@@ -53,9 +53,6 @@ def gen_test_params():
         # Test encoding detection for each file we have of encoding for
         for file_name in listdir(path):
-            ext = splitext(file_name)[1].lower()
-            if ext not in ['.html', '.txt', '.xml', '.srt']:
```
Ah, I made this change (opinionated admittedly) as this filter on suffix tripped me up and I couldn't work out why my tests weren't running.  It was unexpected behaviour to me.  It also appeared that it was functionally meaningless, as it does not actually exclude a single test from running, and so appeared to be a vestigial feature.  Happy to undo however, no biggie.

dan-blanchard replied:
The main reason for it is to deal with hidden files that can crop up on certain OSes.

jpz replied:
Restored :)

jpz commented on chardet/utf1632prober.py:
```diff
@@ -0,0 +1,137 @@
+from .charsetprober import CharSetProber
+
+
+# This class simply looks for occurrences of zero bytes, and infers
```
done, added a few other doc strings where I had comments

jpz commented:
Let me know if you have further things you'd like changed - when you're happy, I'll squash the commits so you can do a clean merge.   (I've left the history there for the moment.)

Naitreey commented:
Hi, guys... What's the status of this patch?

dargueta commented:
Anyone?

jayvdb commented:
ping @jpz , a few nits left.

vikingspy commented:
Did these changes get merged?

dargueta commented:
PR is still open so no

Larrax commented:
Any update on this? I have recently run into several UTF-16 files without BOM and chardet fails horribly with "100% ascii" result. @dan-blanchard / @jpz

jpz commented:
@dan-blanchard I lost my momentum when I was working on this PR a few years ago.  I didn't get any feedback on my PR for about 4 months, and I had moved onto other things at that stage.  I'm sure you were busy with your life, so just explaining how I dropped off the PR.   If you support me making these changes, let me know, I'll be happy to take another look and fixing this up.

Not sure how busy you are maintaining/extending this any more as the repo seems to have a low rate of change, but happy to help out if it's of value.

jpsnyder commented:
I would love to see this get included. This issue has been a real pain point in my project and has forced me to resort to some ugly hacks.

jpsnyder commented:
Bump, any updates on this. This would still be a very valuable change to chardet.

jayvdb commented:
There is now a conflict in `test.py` .  @jpz , any chance you can rebase and fix the whitespace and other simple style changes requested.

terrdavis commented:
I just ran into this issue myself. It would be nice to see these improvements incorporated!

dan-blanchard commented:
This was replaced by #206
